File: backend/src/skills/auth/get_current_user_skill.py
"""GetCurrentUserSkill - Retrieves current authenticated user."""

from sqlmodel import select
from src.models.user import User
from .base_auth_skill import BaseAuthSkill, SkillResult
from .verify_session_skill import VerifySessionSkill


class GetCurrentUserSkill(BaseAuthSkill):
    """
    Skill for retrieving current authenticated user.

    Responsibilities:
    - Verify JWT token
    - Extract user ID from token
    - Fetch user from database
    - Return user object
    """

    def execute(self, token: str) -> SkillResult:
        """
        Get current user from JWT token.

        Args:
            token: JWT access token string

        Returns:
            SkillResult with User object or error message
        """
        try:
            # Verify token and extract payload
            verify_skill = VerifySessionSkill(self.db)
            verify_result = verify_skill.execute(token)

            if not verify_result.success:
                return SkillResult(success=False, error=verify_result.error)

            # Extract user ID from payload
            user_id = verify_result.data.get("user_id")
            if not user_id:
                return SkillResult(success=False, error="User ID not found in token")

            # Fetch user from database
            user = self._get_user_by_id(user_id)
            if not user:
                return SkillResult(success=False, error="User not found")

            return SkillResult(success=True, data=user)

        except Exception as e:
            return SkillResult(success=False, error=f"Failed to get current user: {str(e)}")

    def _get_user_by_id(self, user_id: int) -> User | None:
        """Fetch user by ID from database."""
        statement = select(User).where(User.id == user_id)
        return self.db.exec(statement).first()

File: frontend/lib/skills/get-current-user-skill.ts
/**
 * GetCurrentUserSkill - Retrieves current authenticated user
 *
 * Responsibilities:
 * 1. Get user from Better Auth session (frontend)
 * 2. Optionally fetch user details from backend API
 */

import { BaseAuthSkill, AuthResult } from './base-auth-skill';

interface BetterAuthUser {
  id: string;
  email: string;
  name: string;
}

interface BackendUser {
  id: number;
  email: string;
  created_at: string;
}

interface Session {
  user: BetterAuthUser | null;
}

export class GetCurrentUserSkill extends BaseAuthSkill {
  /**
   * Get current user from Better Auth session
   *
   * @param session - Better Auth session object
   * @returns AuthResult with user data
   */
  async execute(session: Session | null): Promise<AuthResult<BetterAuthUser>> {
    try {
      if (!session || !session.user) {
        return {
          success: false,
          error: 'No active session',
        };
      }

      return {
        success: true,
        data: session.user,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message || 'Failed to get current user',
      };
    }
  }

  /**
   * Fetch user details from backend API
   * Requires valid JWT token
   */
  async fetchBackendUser(): Promise<AuthResult<BackendUser>> {
    try {
      const jwt = this.getJwt();
      if (!jwt) {
        return {
          success: false,
          error: 'No JWT token available',
        };
      }

      const user = await this.callBackendApi<BackendUser>('/api/auth/me', {
        method: 'GET',
        headers: {
          Authorization: `Bearer ${jwt}`,
        },
      });

      return {
        success: true,
        data: user,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message || 'Failed to fetch backend user',
      };
    }
  }
}
